### PR TITLE
[Merged by Bors] - feat: if a function is analytic on a set, its derivative also is, even if the space is not complete

### PR DESCRIPTION
**Pull request by sgouezel**
We already have a version of this theorem, but assuming completeness while this is not necessary: if the function is differentiable, then the power series for its derivative converges, to the given differential (since this is the case in the completion, and the embedding in the completion is an embedding).

This result requires expanding the API around derivatives of analytic functions. As a byproduct, we also write down the derivative of linear maps into multilinear maps (which will be needed for the Faa di Bruno formula).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Calculus/FDeriv/Analytic.lean [hunk 1]
```diff
@@ -361,6 +580,30 @@ protected theorem hasFDerivAt [DecidableEq ι] : HasFDerivAt f (f.linearDeriv x)
   convert f.hasFiniteFPowerSeriesOnBall.hasFDerivAt (y := x) ENNReal.coe_lt_top
   rw [zero_add]
 
+/-- Given `f` a multilinear map, then the derivative of `x ↦ f (g₁ x, ..., gₙ x)` at `x` applied
+to a vector `v` is given by `∑ i, f (g₁ x, ..., g'ᵢ v, ..., gₙ x)`. Version inside a set. -/
+theorem _root_.HasFDerivWithinAt.multilinear_comp
+    [DecidableEq ι] {G : Type*} [NormedAddCommGroup G] [NormedSpace 𝕜 G]
```
Comment (urkud) [inline Mathlib/Analysis/Calculus/FDeriv/Analytic.lean:586]: Do you need `DecidableEq` to state the theorem? If no, then please use `classical` in the proof instead.

Comment (urkud) [reply Mathlib/Analysis/Calculus/FDeriv/Analytic.lean:586]: I see that `ContinuousMultilinearMap.toContinuousLinearMap` uses `DecidableEq`.

File: Mathlib/Analysis/Calculus/FDeriv/Analytic.lean [hunk 2]
```diff
@@ -125,8 +217,17 @@ theorem AnalyticOnNhd.iteratedFDeriv [CompleteSpace F] (h : AnalyticOnNhd 𝕜 f
 @[deprecated (since := "2024-09-26")]
 alias AnalyticOn.iteratedFDeriv := AnalyticOnNhd.iteratedFDeriv
 
+lemma AnalyticOnNhd.hasFTaylorSeriesUpToOn [CompleteSpace F]
+    (n : ℕ∞) (h : AnalyticOnNhd 𝕜 f s) :
+    HasFTaylorSeriesUpToOn n f (ftaylorSeries 𝕜 f) s := by
```
Comment (urkud) [inline Mathlib/Analysis/Calculus/FDeriv/Analytic.lean:222]: Am I right that we need `AnalyticOnNhd`, not `AnalyticOn`, to use `ftaylorSeries`, not `ftaylorSeriesWithin`? If yes, then please add a docstring explaining this. Also, could you please explain in the module docstring which theorems need `[CompleteSpace F]` and which work without this assumption?

Comment (sgouezel) [reply Mathlib/Analysis/Calculus/FDeriv/Analytic.lean:264]: Yes, you're right, I've just explained this. I've also expanded considerably the file-level docstring.

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Pull request closed without merging**